File: C/Stack/Rain_Water.c
/*Name :  Atul Kumar
Github username : atul1510
Repositary name : Algorithms
Problem Description
Program to find the maximum trap water by using stack.
Examples:
 1] INPUT: {1, 1, 5, 2, 1, 8 , 1, 0, 2, 1, 2, 6};
    OUTPUT 31 */
// Implementation in C of the above approach
#include <stdio.h>
// Creating MAX function
#define max(x, y) (((x) > (y)) ? (x) : (y))
// Creating MIN function
#define min(x, y) (((x) < (y)) ? (x) : (y))
// Creating a Function that returns the maximum water
int RainWater(int arr[], int x)
{
	// left[i] contains height of wall
	// left of ith wall
	int left[x];
	// Right [i] contains height of wall
	// right of ith wall
	int right[x];
	// Initialize the result
	int water = 0;
	// Filled the left array
	left[0] = arr[0];
	for (int i = 1; i < x; i++) 
	{
		left[i] = max(left[i - 1], arr[i]);
	}
	// Filled the right array
	right[x - 1] = arr[x - 1];
	for (int i = x - 2; i >= 0; i--) {
		right[i] = max(right[i + 1], arr[i]);
	}
	// Calculate the accumulated water element by element consider the amount of water on i'th bar, the amount of water accumulated on this particular
	// bar will be equal to min(left[i], right[i]) - arr[i]
	for (int i = 1; i < x - 1; i++) {
		int var = min(left[i - 1], right[i + 1]);
		if (var > arr[i]) {
			water += var - arr[i];
		}
	}
	return water;
}
// Main Function
int main()
{
	int arr[] = {1, 1, 5, 2, 1, 8 , 1, 0, 2, 1, 2, 6};
	int m = sizeof(arr) / sizeof(arr[0]);
	printf("%d", RainWater(arr, m));
	return 0;
}

File: Java/Stack/Rain_Water.java
/*Name :  Atul Kumar
Github username : atul1510
Repositary name : Algorithms
Problem Description
Program to find the maximum trap water by using stack .
Examples:
 1] INPUT: {1, 1, 5, 2, 1, 8 , 1, 0, 2, 1, 2, 6};
    OUTPUT 31 */
import java.io.*;
import java.util.*;
// Java implementation of the approach
public class Rain_Water {
	// Function to return the maximum
	// water that can be stored
	public static int RainWater(int[] height)
	{
		// Stores the indices of the bars
		Stack<Integer> stack = new Stack<>();
		// size of the array
		int n = height.length;
		// Stores the final result
		int ans = 0;
		// Loop through the each bar
		for (int i = 0; i < n; i++) {
			// Remove bars from the stack
			// until the condition holds
			while ((!stack.isEmpty())
				&& (height[stack.peek()] < height[i])) {
				// store the height of the top
				// and pop it.
				int pop_height = height[stack.peek()];
				stack.pop();
				// If the stack does not have any
				// bars or the popped bar
				// has no left boundary
				if (stack.isEmpty())
					break;
				// Get the distance between the
				// left and right boundary of
				// popped bar
				int distance = i - stack.peek() - 1;
				// Calculate the min. height
				int min_height
					= Math.min(height[stack.peek()],
							height[i])
					- pop_height;

				ans += distance * min_height;
			}
			// If the stack is either empty or
			// height of the current bar is less than
			// or equal to the top bar of stack
			stack.push(i);
		}
		return ans;
	}
	// Driver code
	public static void main(String[] args)
	{
		int arr[] = {1, 1, 5, 2, 1, 8 , 1, 0, 2, 1, 2, 6};
		System.out.print(RainWater(arr));
	}
}